lib/queue_classic_plus/base.rb

```diff
@@ -75,8 +79,12 @@ def self.can_enqueue?(method, *args)
     end
 
     def self.enqueue(method, *args)
-      if can_enqueue?(method, *args)
-        queue.enqueue(method, *serialized(args))
+      conn = QC.default_conn_adapter.connection
+      conn.transaction do
+        conn.exec("SELECT pg_advisory_xact_lock(#{queue_name_digest})")
```

Comment:
I still have to read how this works internally, but one question that comes to mind is if this might be always creating a lock for every job with a given name, like even when the job name does not have the lock! statement.

Reply:
It will exclusively lock out any other processes trying to add a job to the same queue while the current process is adding a job to the queue. It is not related to the job name.

It does this with or without the `lock!` statement. Note that this "lock" means something different than `lock!`. `lock!` means only 1 instance of a job is allowed to be queued (with the same params).

The DB lock I added here allows `lock!` to work properly by creating a semaphore around the "check if there is a `lock!`, if not enqueue" block.

Reply:
Ahh I see. And what if the enqueuing of a job in a given queue takes quite long, would that make other processes trying to enqueue to have to wait I guess?

Wondering if this could have big performance implications in some cases.

Reply:
The lock is just held while enqueueing, not while the job is running.

There will still be a performance hit as with all locking. I don't think it will be much, but we'll see!